protocol/contracts/zaps/ZapBase.sol

```diff
@@ -0,0 +1,103 @@
+pragma solidity ^0.8.4;
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+abstract contract ZapBase is Ownable {
+  using SafeERC20 for IERC20;
+
+  bool public paused;
+
+  uint256 internal constant DEADLINE = 0xf000000000000000000000000000000000000000000000000000000000000000;
+
+  // fromToken => swapTarget (per curve, univ2 and balancer) approval status
+  mapping(address => mapping(address => bool)) public approvedTargets;
+
+  receive() external payable {
+    require(msg.sender != tx.origin, "Do not send ETH directly");
+  }
+
+  /**
+    @dev Adds or removes an approved swapTarget
+    * swapTargets should be Zaps and must not be tokens!
+    @param _tokens An array of tokens
+    @param _targets An array of addresses of approved swapTargets
+    @param _isApproved An array of booleans if target is approved or not
+    */
+  function setApprovedTargets(
+    address[] calldata _tokens,
+    address[] calldata _targets,
+    bool[] calldata _isApproved
+  ) external onlyOwner {
+    uint256 _length = _isApproved.length;
+    require(_targets.length == _length && _tokens.length == _length, "Invalid Input length");
+
+    for (uint256 i = 0; i < _length; i++) {
+      approvedTargets[_tokens[i]][_targets[i]] = _isApproved[i];
+    }
+  }
+
+  /**
+    @dev Toggles the contract's active state
+     */
+  function toggleContractActive() external onlyOwner {
+    paused = !paused;
+  }
+
+  /**
+    @notice Approve a token for spending with finite allowance
+    @param _token The ERC20 token to approve
+    @param _spender The spender of the token
+    @param _amount The allowance to grant to the spender
+     */
+  function _approveToken(
```

Comment:
Is this required as opposed to adding in-line where required? extra gas for a pretty small bit of syntactic sugar?

Reply:
yh good call. there were some cases i had to use inline for stack management.